Notifications API: default silent to platform convention

marcoscaceres opened the pull request:
#### bb5a5fbce2e993ee60fbb3046b6e6c03693b49a0
<pre>
Notifications API: default silent to platform convention
<a href="https://bugs.webkit.org/show_bug.cgi?id=256828">https://bugs.webkit.org/show_bug.cgi?id=256828</a>
rdar://109390045

Reviewed by Brady Eidson.

Makes the silent member on the NotificationOption dictionary default to null.
This is then reflected by the silent attribute of the Notification interface.

Spec change:
<a href="https://github.com/whatwg/notifications/pull/194">https://github.com/whatwg/notifications/pull/194</a>

* LayoutTests/imported/w3c/web-platform-tests/interfaces/notifications.idl:
* LayoutTests/imported/w3c/web-platform-tests/notifications/constructor-basic.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/notifications/constructor-basic.https.html:
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/Notification.idl:
* Source/WebCore/Modules/notifications/NotificationOptions.idl:

Canonical link: <a href="https://commits.webkit.org/264397@main">https://commits.webkit.org/264397@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/261808eeb5fa2bee591d8253427b14778130f4c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7133 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7380 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7559 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8749 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7357 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7143 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10255 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7917 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8855 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5272 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6490 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14235 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6934 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9464 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7069 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5781 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6399 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1780 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6812 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on LayoutTests/imported/w3c/web-platform-tests/notifications/constructor-basic.https.html:
```diff
@@ -18,4 +18,22 @@
+}, "Constructing a notification without a NotificationOptions defaults to null");
+
+test(() => {
+    for (const expected of [null, true, false]) {
```
It would be good to also try undefined here and maybe some others such as a number of a string.

marcoscaceres replied:
Good idea. Added.

annevk commented on Source/WebCore/Modules/notifications/Notification.h:
```diff
@@ -88,7 +88,7 @@ class Notification final : public RefCounted<Notification>, public ActiveDOMObje
     const URL& icon() const { return m_icon; }
     JSC::JSValue dataForBindings(JSC::JSGlobalObject&);
-    bool silent() const { return m_silent == std::nullopt ? false : *m_silent; }
+    const std::optional<bool> silent() const { return m_silent; }
```
I'm surprised we don't use `Markable<>` for our IDL stuff. cc @mdubet

mdubet replied:
`Markable` doesn't work with bool because we do need the 3 states (null, false, true) and we can't encode them on a single bit.

mdubet replied:
The first `const` (on the return value) seems unnecessary: `std::optional<bool> silent() const { return m_silent; }`

marcoscaceres replied:
oops, always get those mixed up.

annevk reviewed (commented):
Non-reviewer looks good!

marcoscaceres committed:
Notifications API: default silent to platform convention
https://bugs.webkit.org/show_bug.cgi?id=256828
rdar://109390045

Reviewed by Brady Eidson.

Makes the silent member on the NotificationOption dictionary default to null.
This is then reflected by the silent attribute of the Notification interface.

Spec change:
https://github.com/whatwg/notifications/pull/194

* LayoutTests/imported/w3c/web-platform-tests/interfaces/notifications.idl:
* LayoutTests/imported/w3c/web-platform-tests/notifications/constructor-basic.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/notifications/constructor-basic.https.html:
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/Notification.idl:
* Source/WebCore/Modules/notifications/NotificationOptions.idl:

Canonical link: https://commits.webkit.org/264397@main
